src/dragonball/src/vm/mod.rs

```diff
@@ -89,8 +144,659 @@ impl Default for VmConfigInfo {
+        };
+        instance_state == InstanceState::Running
+    }
+
+    /// Save VM instance exit state
+    pub fn vm_exit(&self, exit_code: i32) {
+        if let Ok(mut info) = self.shared_info.write() {
+            info.state = InstanceState::Exited(exit_code);
+        } else {
+            error!(
+                self.logger,
+                "Failed to save exit state, couldn't be written due to poisoned lock"
+            );
+        }
+    }
+
+    /// Create device operation context.
+    /// vm is not running, return false
+    /// vm is running, but hotplug feature is not enable, return error
+    /// vm is running, but upcall initialize failed, return error
+    /// vm is running, upcall initialize OK, return true
+    pub fn create_device_op_context(
+        &mut self,
+        epoll_mgr: Option<EpollManager>,
+    ) -> std::result::Result<DeviceOpContext, StartMicroVmError> {
+        if !self.is_vm_initialized() {
+            Ok(DeviceOpContext::create_boot_ctx(self, epoll_mgr))
+        } else {
+            self.create_device_hotplug_context(epoll_mgr)
+        }
+    }
+
+    pub(crate) fn check_health(&self) -> std::result::Result<(), StartMicroVmError> {
+        if self.kernel_config.is_none() {
+            return Err(StartMicroVmError::MissingKernelConfig);
+        }
+        Ok(())
+    }
+
+    pub(crate) fn get_dragonball_info(&self) -> (String, String) {
+        let guard = self.shared_info.read().unwrap();
+        let instance_id = guard.id.clone();
+        let dragonball_version = guard.vmm_version.clone();
+
+        (dragonball_version, instance_id)
+    }
+}
+
+impl Vm {
+    pub(crate) fn init_vcpu_manager(
+        &mut self,
+        vm_as: GuestAddressSpaceImpl,
+        vcpu_seccomp_filter: BpfProgram,
+    ) -> std::result::Result<(), VcpuManagerError> {
+        let vcpu_manager = VcpuManager::new(
+            self.vm_fd.clone(),
+            &self.kvm,
+            &self.vm_config,
+            vm_as,
+            vcpu_seccomp_filter,
+            self.shared_info.clone(),
+            self.device_manager.io_manager(),
+            self.epoll_manager.clone(),
+        )?;
+        self.vcpu_manager = Some(vcpu_manager);
+
+        Ok(())
+    }
+
+    /// get the cpu manager's reference
+    pub(crate) fn vcpu_manager(
+        &self,
+    ) -> std::result::Result<std::sync::MutexGuard<'_, VcpuManager>, VcpuManagerError> {
+        self.vcpu_manager
+            .as_ref()
+            .ok_or(VcpuManagerError::VcpuManagerNotInitialized)
+            .map(|mgr| mgr.lock().unwrap())
+    }
+
+    /// Pause all vcpus and record the instance downtime
+    pub fn pause_all_vcpus_with_downtime(&mut self) -> std::result::Result<(), VcpuManagerError> {
+        let ts = TimestampUs::default();
+        self.start_instance_downtime = ts.time_us;
+
+        self.vcpu_manager()?.pause_all_vcpus()?;
+
+        Ok(())
+    }
+
+    /// Resume all vcpus and calc the intance downtime
```

Comment:
- Nit: typo: "instance"
- The start of the downtime timestamp is recorded in `pause_all_vcpus_with_downtime()` but I don't see where the overall downtime calculation is...?

Reply:
Added.